### PR TITLE
Fixed visibility logic on Price field options.

### PR DESCRIPTION
**Pull request by agilewarealok**
Steps to reproduce the issue
----------------------------------------

1. Create a new Price Set
2. Add a Price field (Radio) with Visibility: Admin and 2 options.
3. Edit one of the options and change its visibility to Public
4. Add one option to that field with Visibility: Admin

Error is displayed: "You must choose 'Admin' visibility for this price option, as it belongs to a field with 'Admin' visibility.

Expected Behavior
----------------------------------------

1. Public Price field should at  least have one public option.
2. Admin Price field must contain only admin options.

Before
----------------------------------------
The current implementation was doing above checks without considering the new option which is being added. 

After
----------------------------------------
Fixed the implementation and it is now working as expected. 

Comments
----------------------------------------
_Agileware Ref: CIVICRM-1168_

### CODE REVIEW

**Comment by seamuslee001** [comment]
ping @lcdservices i believe you have run into similar issues to this

**Comment by eileenmcnaughton** [comment]
Yeah I was also wondering if @agh1 or @Stoob had an opinion on this

**Comment by Stoob** [comment]
I don't necessarily agree with expected behavior number 2 would like clarification if possible please.

Use case possibility: Would a user have to delve also into the options within each newly switched-to-admin field and switch each option to admin too, in order to prevent this error?

Is it not sufficient to simply assume all a admin-only field's options within as also admin-only, and remove this warning/precaution altogether?  This is akin to a custom field group - if the field group is not public, none of the fields within will be visible either.

**Comment by agh1** [comment]
I agree with @stoob that a public option makes no sense if the whole field is admin-only.  Validation seems too late in the game: the admin option should be forced/frozen when you go to set up the option.

However, the change in this PR (as I read the code) makes sense: if there are no public options on a public field, you shouldn't be stopped from creating a new option, so long as the new option is public.  If you are creating a new option for an admin-only field, you shouldn't be stopped just because some other option is public.

**Comment by Stoob** [comment]
Good point.  And mea culpa I'm haven't read the code.  

I guess I'm just wondering why Civi is validating creation of the public/admin setting of options at all?  I mean if you *want* to create an admin-only option for a pubic field, you should be allowed to do so.  That would mean the field is public but certain admin-only-options within aren't.   

If you want to create a public option for an admin-only field, you should be allowed to do that too.  The field won't show up for the public though, because the field itself is set to admin-only.  That's easy enough to fix and understand why.

**Comment by lcdservices** [comment]
I tend to agree with @Stoob's last comments. I'm not sure we even need validation along these lines when creating fields and options. It's not unusual for organizations to flip visibility at times in order to manage particular temporary use cases. As long as the visibility is respected in the frontend, I'd lean toward removing complex validation logic in the backend.

**Comment by jusfreeman** [comment]
> As long as the visibility is respected in the frontend, I'd lean toward removing complex validation logic in the backend.

@lcdservices yes, I agree with that point. Any volunteers to submit a PR to remove the backend's complex validation logic?

Until then, can we at least fix the current situation - because it's broken.

**Comment by Stoob** [comment]
@jusfreeman shrug, I'm not a code so can't submit a PR but do support removing the validation for options.  If there is an effort to do so I will support it financially.  In the meantime will look to others for merge decision.

**Comment by mattwire** [comment]
@jusfreeman @Stoob Reading this, if we removed the validation options that simplifies the code and also means this PR would not be needed?  Is that right?  
As it is, I feel like a change to the validation logic would require a unit test, whereas simplifying the code and removing the validation logic would not - and we'd end up with simpler, less bug-prone code.
If that's correct, I'd prefer to see the "simplification" PR so we only have to do one review.

**Comment by agh1** [comment]
That sounds fine so long as the price set on a public form suppresses a
public field with only admin options and an admin field even if it has
public options.  I'm not sure if the first happens.

Unless that can be confirmed and the validation removed quickly and
cleanly, I'd be in favor of adopting this PR because it is a
straightforward resolution of a clear-cut bug and it's consistent with the
logic of the code as it stands.  Removing validation would be a change in
intended behavior, so I'd be less eager to adopt that hastily.

**Comment by stesi561** [comment]
Just a note to say that while I agree simplification would be good in the long term, given this is currently broken - and I imagine quite confusing for anyone who stumbles across it, I think it would be a good idea to adopt this PR.

**Comment by jusfreeman** [comment]
@mattwire at this stage, can we just fix this bug using PR? It's a really confounding bug that's pretty simple to fix.

Agree with simplifying the code later on in a separate PR.

**Review by agh1 — APPROVED** [review]
* General standards
    * ([`r-explain`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-explain)) __PASS__: This is reasonable, but it doesn't quite cover the overall context.
    * ([`r-user`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-user)) __PASS__: While CiviCRM will still let you create messed-up field/option visibility combinations, this change now allows you to resolve this by editing an option to agree with the field.
    * ([`r-doc`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-doc)) __PASS__
    * ([`r-run`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-run)) __PASS__: see below
* Developer standards
    * ([`r-tech`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-tech)) __PASS__: The change preserves compatibility with existing callers/code/downstream.
    * ([`r-code`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-code)) __PASS__: The functionality, purpose, and style of the code seems clear+sensible.
    * ([`r-maint`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-maint)) __PASS__: The change does not add test coverage, but any proper test coverage for this issue would fail in other cases.
    * ([`r-test`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-test)) __PASS__: The test results are all-clear.

As we can see from the discussion, the validation on price field/option visibility combinations is really spotty.  I opened [dev/core#1040](https://lab.civicrm.org/dev/core/issues/1040) to hold further discussion on this: this PR is an improvement, but it barely scratches the surface of problems on this.

I tested the following procedure and results on master first:

1. Created "test master price set" for events with event fee.

2. Create Field A (public) with option 1 (public) and option 2 (admin).  No problem.

3. Create Field B (public) with option 1 (admin) and option 2 (admin).  Validation error on all price options and on Visibility field (first screenshot).  Edit option 2 to public, and it works.

4. Edit Field B (public) option 2 (public) and set it to admin.  Validation error on Visibility field.  Set option back to public, and it works.

5. Delete Field B (public) option 2 (public), and it works, leaving only option 1 (private).  This is undesirable: a public field with only admin options.

6. Create new option 2 (public) in field B (public).  Validation error on Visibility field.  This is a bad problem: you **want** the price option to be public, and it is preventing you from doing what it asks.

7. Edit Field B (public, with only an admin option) and set it to admin, and it works.

8. Edit Field A (public, with a public option and an admin option) and set it to admin, and it works.

9. Create Field C (public) with option 1 (public) and option 2 (public).  No problem.

10. Edit Field C (public, with two public options) and set it to admin, and it works.  This is a little confusing: an admin field with only public options.

11. Edit Field C (admin) option 1 (public) and set it to admin.  Validation error on Visibility field.  This is a bad problem: you **want** the price option to be admin, and it is preventing you from doing what it asks.

12. Create Field D (admin) with option 1 (public) and option 2 (public).  No problem.  This is a little confusing: an admin field with only public options.

13. Delete Field A (admin) option 2 (admin).  No problem.  This is a little confusing: an admin field with only public options.

14. Edit Field B (admin) option 1 (admin) and set it to public.  Validation error on Visibility field.

15. Edit Field B (admin, with only an admin option) and set it to public.  No problem.  This is undesirable: a public field with only an admin option.

Here's a summary, with undesirable results in italics:

|                | Admin w/ only public options | Public w/ only admin options |
| -------------- | ---------------------------- | ---------------------------- |
| Creating:      | *Allowed (12)*               | Not allowed (3)              |
| Delete option: | *Allowed (13)*               | *Allowed (5)*                |
| Change option: | Not allowed (14)             | Not allowed (4)              |
| Fix:           | *Not allowed (11)*           | *Not allowed (6)*            |
| Edit field:    | *Allowed (10)*               | *Allowed (15)*               |

I then tried the same procedure with this PR:

1. Created "test master price set" for events with event fee.

2. Create Field A (public) with option 1 (public) and option 2 (admin).  No problem.

3. Create Field B (public) with option 1 (admin) and option 2 (admin).  Validation error on all price options and on Visibility field (first screenshot).  Edit option 2 to public, and it works.

4. Edit Field B (public) option 2 (public) and set it to admin.  Validation error on Visibility field.

5. Delete Field B (public) option 2 (public), and it works, leaving only option 1 (private).  No problem.

6. Create new option 2 (public) in field B (public). No problem.  (Deleted the option to keep consistent with above workflow.)

7. Edit Field B (public, with only an admin option) and set it to admin, and it works.

8. Edit Field A (public, with a public option and an admin option) and set it to admin.  No problem.

9. Create Field C (public) with option 1 (public) and option 2 (public).

10. Edit Field C (public, with two public options) and set it to admin.  This is a little confusing: an admin field with only public options.

11. Edit Field C (admin) option 1 (public) and set it to admin.  No problem.

12. Create Field D (admin) with option 1 (public) and option 2 (public).  No problem.  This is a little confusing: an admin field with only public options.

13. Delete Field A (admin) option 2 (admin).  No problem.  This is a little confusing: an admin field with only public options.

14. Edit Field B (admin) option 1 (admin) and set it to public.  Validation error on Visibility field.

15. Edit Field B (admin, with only an admin option) and set it to public.  No problem.  This is undesirable: a public field with only an admin option.

In summary, with undesirable results in italics and changes in bold:

|                | Admin w/ only public options | Public w/ only admin options |
| -------------- | ---------------------------- | ---------------------------- |
| Creating:      | *Allowed (12)*               | Not allowed (3)              |
| Delete option: | *Allowed (13)*               | *Allowed (5)*                |
| Change option: | Now allowed (14)             | Not allowed (4)              |
| Fix:           | **Allowed (11)**             | **Allowed (6)**              |
| Edit field:    | *Allowed (10)*               | *Allowed (15)*               |

**Comment by eileenmcnaughton** [comment]
Merging based on (thorough) review by @agh1
